### PR TITLE
Favorites tab for the construction menu

### PR DESCRIPTION
**Pull request by Arteben**
## About the PR
This PR adds a new 'Favorites' tab where you can add recipes by collecting them from various tabs into one for your convenience. A button is added in the top right corner to delete or add recipes. Additionally, there's a slight change in resetting text from the search bar. When a new tab is selected, the text and, accordingly, the search are reset.

## Technical details

Simple changes, only for the  construction window UI.

## Media

image

https://github.com/space-wizards/space-station-14/assets/2144635/fc54c2ca-5548-4673-be44-0d61d50dc227

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added favorite tab for construction!

### CODE REVIEW

**Comment by EmoGarbage404** [comment]
it's kind of annoying that it doesn't make any attempt to save the actual favorites list across sessions.

**Comment by Arteben** [comment]
> it's kind of annoying that it doesn't make any attempt to save the actual favorites list across sessions.

I felt like there wasn't enough favorites, even within one session. I've implemented them. It might be possible between sessions, but for now, it's too difficult for me.

File: Content.Client/Construction/UI/ConstructionMenu.xaml [hunk 1]
```diff
@@ -1,29 +1,35 @@
...
-                <BoxContainer Orientation="Horizontal" Align="Center">
-                    <TextureRect Name="TargetTexture" HorizontalAlignment="Right" Stretch="Keep"/>
-                    <Control MinSize="10 0"/>
-                    <BoxContainer Orientation="Vertical">
-                        <RichTextLabel Name="TargetName"/>
-                        <RichTextLabel Name="TargetDesc"/>
+                <BoxContainer Orientation="Vertical" HorizontalExpand="True" Margin="0 0 0 5">
+                    <BoxContainer Orientation="Horizontal" Align="Center">
+                        <TextureRect Name="TargetTexture" HorizontalAlignment="Right" Stretch="Keep"/>
+                        <Control MinSize="10 0"/>
```
Comment (metalgearsloth) [inline Content.Client/Construction/UI/ConstructionMenu.xaml:18]: Should this dummy control just be a margin instead

File: Content.Client/Construction/UI/ConstructionMenuPresenter.cs [hunk 2]
```diff
@@ -185,40 +205,50 @@ private void OnViewPopulateRecipes(object? sender, (string search, string catago
...
             foreach (var prototype in _prototypeManager.EnumeratePrototypes<ConstructionPrototype>())
             {
                 var category = prototype.Category;
 
                 if (!string.IsNullOrEmpty(category))
                     uniqueCategories.Add(category);
             }
+            var sortedCategories = uniqueCategories.OrderBy(x => Loc.GetString(x)).ToList();
+            // hard-coded to show all recipes
+            sortedCategories.Insert(0, _forAllCategoryName);
```
Comment (metalgearsloth) [inline Content.Client/Construction/UI/ConstructionMenuPresenter.cs:222]: This is going to shift the entire list along which is incredibly slow, just allocate sortedcategories up front (you already have its size) then add _forAllCategoryName then add the localised string afterwards.

File: Content.Client/Construction/UI/ConstructionMenuPresenter.cs [hunk 3]
```diff
@@ -185,40 +205,50 @@ private void OnViewPopulateRecipes(object? sender, (string search, string catago
...
+            if (_favoritedRecipes.Count > 0)
+            {
+                // hard-coded to show favorites if it need
+                sortedCategories.Insert(1, _favoriteCatName);
+            }
 
-            var array = uniqueCategories.OrderBy(Loc.GetString).ToArray();
-            Array.Sort(array);
+            var categoriesArray = sortedCategories.ToArray();
+            _constructionView.Category.Clear();
```
Comment (metalgearsloth) [inline Content.Client/Construction/UI/ConstructionMenuPresenter.cs:230]: This seems entirely unnecessary when you can just make sortedCategories an array?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Client/Construction/UI/ConstructionMenuPresenter.cs (hunk 2 above)
Comment (Arteben) [reply Content.Client/Construction/UI/ConstructionMenuPresenter.cs:222]: I don't know, this item was already there, and I didn't touch it. But I'll try to remove it.

File: Content.Client/Construction/UI/ConstructionMenu.xaml (hunk 1 above)
Comment (Arteben) [reply Content.Client/Construction/UI/ConstructionMenu.xaml:18]: I don't know, this item was already there, and I didn't touch it. But I'll try to remove it.

Comment (Arteben) [reply Content.Client/Construction/UI/ConstructionMenu.xaml]: deleted  these controls

File: Content.Client/Construction/UI/ConstructionMenuPresenter.cs (hunk 2 above)
Comment (Arteben) [reply Content.Client/Construction/UI/ConstructionMenuPresenter.cs]: I removed the insertions of elements into the lists and tried to simplify this code. Maybe you had something else in mind, but this is what I came up with.

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**